### PR TITLE
EOS-26608: config, cfgen support nodes without motr srvces and devices

### PR DESCRIPTION
**Pull request by suvratjoshi**
Currently Hare 'config' stage is assuming that each node / pod will
contain motr services and devices mentioned in CDF file.
In any scenarios, config stage of hare should check, if motr services
are there and then only generate motr related fids.
In case motr services are not there, do not start/intiate motr
dependent services. Also currently 'config' is failing in s3server pod,
if 'motr-hare_keys.json' is not present, this needs to be fixed for
s3server pod.

Signed-off-by: Suvrat Joshi <suvrat.joshi@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: cfgen/cfgen [hunk 1]
```diff
@@ -404,6 +404,9 @@ def enrich_cluster_desc(desc: Dict[str, Any], mock_p: bool) -> None:
         if 'm0_servers' not in node:
             continue
 
+        m0d = node['m0_servers']
```
Comment (knekrasov) [inline cfgen/cfgen:407]: If there is no `m0_servers` key in this dict, `KeyError` will be thrown immediately (so consequent check for `is None` doesn't make sense).

https://stackoverflow.com/questions/11041405/why-dict-getkey-instead-of-dictkey

File: cfgen/cfgen [hunk 2]
```diff
@@ -460,6 +463,9 @@ def validate_nodes_desc(nodes_desc: List[Dict[str, Any]]) -> None:
         _assert(node['facts'][ipaddr_key(iface)],
                 f"""{name}: {iface!r} interface has no IP address
 Make sure the value of data_iface in the CDF is correct.""")
+        m0d = node['m0_servers']
+        if m0d is None:
```
Comment (knekrasov) [inline cfgen/cfgen:467]: The same issue with square brackets and KeyError.

File: cfgen/cfgen [hunk 3]
```diff
@@ -2420,21 +2426,26 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
         node_id = ConfNode.build(conf, root_id, node)
         encl_id = ConfEnclosure.build(conf, rack_id, node_id)
         ctrl_ids: List[Tuple[Optional[Oid], Any]] = []
-        for m0d in node.get('m0_servers', []):
-            if m0d['_io_disks']:
-                ctrl_ids.append((ConfController.build(conf, encl_id), m0d))
-            else:
-                ctrl_ids.append((None, m0d))
+        m0srv = node['m0_servers']
```
Comment (knekrasov) [inline cfgen/cfgen:2429]: ..and here also

File: cfgen/cfgen [hunk 4]
```diff
@@ -2420,21 +2426,26 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
...
 
-        confd_p = False
-        for ctrl_id in ctrl_ids:
-            m0d = ctrl_id[1]
-            conf_ctrl_id: Optional[Oid] = ctrl_id[0]
-            ConfProcess.build(conf, node_id, node, ProcT.m0_server, m0d,
-                              conf_ctrl_id)
-            if m0d['runs_confd']:
-                confd_p = True
+        m0srv = node['m0_servers']
```
Comment (knekrasov) [inline cfgen/cfgen:2439]: .. and here

File: hax/hax/queue/cli.py [hunk 5]
```diff
@@ -49,7 +46,6 @@ def parse_opts(ctx, queue: str, type: str, payload: str):
 
 def main():
     _setup_logging()
-    inject.configure(di_configuration)
```
Comment (knekrasov) [inline hax/hax/queue/cli.py]: Why do you remove this?

File: provisioning/miniprov/hare_mp/main.py [hunk 6]
```diff
@@ -288,18 +287,31 @@ def _start_hax(utils: Utils,
...
+    try:
+        if checkRpm('cortx-motr') != 0:
+            logging.error('\'cortx-motr\' is not installed')
+            exit(-1)
+        if checkRpm('consul') != 0:
+            logging.error('\'consul\' is not installed')
+            exit(-1)
+        if checkRpm('cortx-hare') != 0:
+            logging.error('\'cortx-hare\' is not installed')
+            exit(-1)
```
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:299]: It seems like you're reverting some important changes in your commit. 
Can you make your changes minimal? I believe this reverting is not required by your PR (and even if it is, then such a change deserves a separate commit).

**Review by knekrasov — CHANGES_REQUESTED** [review]

**Comment by mssawant** [comment]
>Currently Hare 'config' stage is assuming that each node / pod will
contain motr services and devices mentioned in CDF file.
In any scenarios, config stage of hare should check, if motr services
are there and then only generate motr related fids.
In case motr services are not there, do not start/intiate motr
dependent services. Also currently 'config' is failing in s3server pod,
if 'motr-hare_keys.json' is not present, this needs to be fixed for
s3server pod.

@suvratjoshi, Idea is to have same CDF on all the nodes, I am confused by the statement `In any scenarios, config stage of hare should check, if motr services are there and then only generate motr related fids.`, does motr services also include s3server? 
So,
- CDF should be same on motr ioservices pod and s3server pods
- consul-client-conf for s3server pod must contain only the s3server services.
- while the motr ioservice pod will not have s3services present
If we take care of the 1, 2 and 3 must already be supported.

File: provisioning/miniprov/hare_mp/cdf.py [hunk 7]
```diff
@@ -330,57 +332,64 @@ def _get_data_devices(self, machine_id: str, cvg: int) -> DList[Text]:
...
-        except TypeError:
-            no_m0clients = 2
-        # Currently, there is 1 m0d per cvg.
-        # We will create 1 IO service entry in CDF per cvg.
-        # An IO service entry will use data devices from corresponding cvg.
-        # meta data device is taken from motr-hare shared store.
-        servers = DList([
-            M0ServerDesc(
+        servers = None
+        no_m0clients = 0
```
Comment (mssawant) [inline provisioning/miniprov/hare_mp/cdf.py:356]: We are hard coding this, same for the s3 instances, We need to request these keys, cortx>motr>client_instances and cortx>s3>service_instances to move under the node hierarchy.
Or we need to check if a node is running S3 services then use cortx>s3>service_instances and create those many number of `m0_clients` instances similar to `m0_servers`
e.g.
```
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2            # max number of other Motr clients this host may have
  - hostname: client-21    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-22    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16
```
These are some pure client nodes.

**Comment by mssawant** [comment]
@suvratjoshi, this is how the CDF must look like with pure client nodes, in our case with the pods that run only s3servers, you need to create only m0_clients entry for that node in CDF.
```
# Cluster Description File (CDF).
# See `cfgen --help-schema` for the format description.

nodes:
  - hostname: sage-tier1-1 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/nvme0n1
            - /dev/nvme1n1
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier1-2 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/nvme0n1
            - /dev/nvme1n1
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier2-1 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/disk/by-id/wwn-0x5000c500301706d3
            #- /dev/disk/by-id/wwn-0x5000c500301706df
            - /dev/disk/by-id/wwn-0x5000c5003017072f
            - /dev/disk/by-id/wwn-0x5000c5003016d65b
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier2-2 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/disk/by-id/wwn-0x5000c5003016d6d3
            - /dev/disk/by-id/wwn-0x5000c5003016d653
            - /dev/disk/by-id/wwn-0x5000c500301706db
            - /dev/disk/by-id/wwn-0x5000c5003016d5c3
            - /dev/disk/by-id/wwn-0x5000c50030170723
            - /dev/disk/by-id/wwn-0x5000c500301706d7
            - /dev/disk/by-id/wwn-0x5000c500301706f7
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier2-3 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
#            - /dev/disk/by-id/wwn-0x5000c5003016d66b
            - /dev/disk/by-id/wwn-0x5000c500301706e3
            - /dev/disk/by-id/wwn-0x5000c5003017071b
            - /dev/disk/by-id/wwn-0x5000c5003017075f
            - /dev/disk/by-id/wwn-0x5000c50030170717
            - /dev/disk/by-id/wwn-0x5000c5003017075b
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier2-4 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/disk/by-id/wwn-0x5000c5003016d68f
            - /dev/disk/by-id/wwn-0x5000c5003016d5d3
            - /dev/disk/by-id/wwn-0x5000c5003016d673
            - /dev/disk/by-id/wwn-0x5000c500301706ef
            - /dev/disk/by-id/wwn-0x5000c5003017073b
            - /dev/disk/by-id/wwn-0x5000c50030170737
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2             # max number of other Motr clients this host may have
  - hostname: sage-tier3-1 # [user@]hostname
    data_iface: ib0        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
                           # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/disk/by-id/wwn-0x5000c50095bd089b
            - /dev/disk/by-id/wwn-0x5000c50095bd092f
            - /dev/disk/by-id/wwn-0x5000c50095bd0d0b
            - /dev/disk/by-id/wwn-0x5000c50095bd0f97
            - /dev/disk/by-id/wwn-0x5000c50095bd0ffb
            - /dev/disk/by-id/wwn-0x5000c50095bd210f
            - /dev/disk/by-id/wwn-0x5000c50095bd221b
            - /dev/disk/by-id/wwn-0x5000c50095bd2237
            - /dev/disk/by-id/wwn-0x5000c50095bd2dc3
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 2            # max number of other Motr clients this host may have
  - hostname: client-21    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-22    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-23    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-24    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-25    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-26    # [user@]hostname
    data_iface: ib1        # name of data network interface
    data_iface_type: o2ib  # type of network interface (optional);
    m0_clients:
      s3: 0                # number of S3 servers to start
      other: 16            # max number of other Motr clients this host may have
  - hostname: client-27    # [user@]hostname
    data_iface: ib1        # name of data network interface
pools:
  - name: tier1-nvme-p1
    #type: sns  # optional; supported values: "sns" (default), "dix", "md"
    disk_refs:
      - { path: /dev/nvme0n1, node: sage-tier1-1 }
      - { path: /dev/nvme1n1, node: sage-tier1-1 }
      - { path: /dev/nvme0n1, node: sage-tier1-2 }
      - { path: /dev/nvme1n1, node: sage-tier1-2 }
    data_units: 2
    parity_units: 1
    allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 1 }
  - name: tier2-ssd-p2
    disk_refs:
      - { path: /dev/disk/by-id/wwn-0x5000c500301706d3, node: sage-tier2-1 }
      - { path: /dev/disk/by-id/wwn-0x5000c5003017072f, node: sage-tier2-1 }
#      - { path: /dev/disk/by-id/wwn-0x5000c5003016d6d3, node: sage-tier2-2 }
#      - { path: /dev/disk/by-id/wwn-0x5000c5003016d653, node: sage-tier2-2 }
      - { path: /dev/disk/by-id/wwn-0x5000c5003016d68f, node: sage-tier2-4 }
      - { path: /dev/disk/by-id/wwn-0x5000c5003016d5d3, node: sage-tier2-4 }
    data_units: 2
    parity_units: 1
    allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 1 }
  - name: tier3-hdd-p3
    disk_refs:
      - { path: /dev/disk/by-id/wwn-0x5000c50095bd089b, node: sage-tier3-1 }
      - { path: /dev/disk/by-id/wwn-0x5000c50095bd092f, node: sage-tier3-1 }
      - { path: /dev/disk/by-id/wwn-0x5000c50095bd0d0b, node: sage-tier3-1 }
      - { path: /dev/disk/by-id/wwn-0x5000c50095bd0f97, node: sage-tier3-1 }
    data_units: 2
    parity_units: 1
    allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 1 }
...
    data_units: 8
    parity_units: 2
    allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 2 }
profiles:
  - name: prof-0
    pools: [ tier1-nvme-p1, tier2-ssd-p2, tier3-hdd-p3, tier2-ssd, tier3-hdd ]
```

**Comment by mssawant** [comment]
@suvratjoshi, can you post a CDF generated using this patch. cc @vaibhavparatwar.

File: provisioning/miniprov/hare_mp/cdf.py (hunk 7 above)
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/cdf.py:356]: We are already reading the confstore keys and returning values for "no_m0clients" and "nr_s3_instances".
"if(self.utils.is_motr_component(machine_id))" This check is reading confstore key to decide whether motr component is present or not.

Only the reason I had to put variable inits to some values here was, to have non-none values as the returning class "NodeDesc" has fixed set of parameters.

**Comment by suvratjoshi** [comment]
> @suvratjoshi, can you post a CDF generated using this patch. cc @vaibhavparatwar.

```
pools:
- allowed_failures:
    disk: 0
    ctrl: 0
    encl: 0
    rack: 0
    site: 0
  spare_units: 0
  name: storage-set-1__sns
  data_units: 1
  parity_units: 0
  disk_refs:
  - path: /dev/sdc
    node: data-node1
  - path: /dev/sdd
    node: data-node1
  - path: /dev/sdf
    node: data-node1
  - path: /dev/sdg
    node: data-node1
  - path: /dev/sdc
    node: data-node2
  - path: /dev/sdd
    node: data-node2
  - path: /dev/sdf
    node: data-node2
  - path: /dev/sdg
    node: data-node2
  - path: /dev/sdc
    node: data-node3
  - path: /dev/sdd
    node: data-node3
  - path: /dev/sdf
    node: data-node3
  - path: /dev/sdg
    node: data-node3
  type: sns
- allowed_failures:
    disk: 0
    ctrl: 0
    encl: 0
    rack: 0
    site: 0
  spare_units: 0
  name: storage-set-1__dix
  data_units: 1
  parity_units: 0
  disk_refs:
  - path: /dev/sdb
    node: data-node1
  - path: /dev/sde
    node: data-node1
  - path: /dev/sdb
    node: data-node2
  - path: /dev/sde
    node: data-node2
  - path: /dev/sdb
    node: data-node3
  - path: /dev/sde
    node: data-node3
  type: dix
create_aux: false
profiles:
- pools:
  - storage-set-1__sns
  - storage-set-1__dix
  name: Profile_the_pool
fdmi_filters: null
nodes:
- m0_servers:
  - io_disks:
      data:
      - path: /dev/sdc
        size: 53687091200
        blksize: 4096
      - path: /dev/sdd
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sdb'']'
    runs_confd: false
  - io_disks:
      data:
      - path: /dev/sdf
        size: 53687091200
        blksize: 4096
      - path: /dev/sdg
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sde'']'
    runs_confd: false
  - io_disks:
      data: []
      meta_data: null
    runs_confd: true
  data_iface: dummy
  hostname: data-node1
  data_iface_ip_addr: data-node1
  m0_clients:
    other: 0
    s3: 0
  memorysize_mb: 15866.52734375
  machine_id: aaa120a9e051d103c164f605615c32a4
  processorcount: 8
  data_iface_type: tcp
- m0_servers:
  - io_disks:
      data:
      - path: /dev/sdc
        size: 53687091200
        blksize: 4096
      - path: /dev/sdd
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sdb'']'
    runs_confd: false
  - io_disks:
      data:
      - path: /dev/sdf
        size: 53687091200
        blksize: 4096
      - path: /dev/sdg
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sde'']'
    runs_confd: false
  - io_disks:
      data: []
      meta_data: null
    runs_confd: true
  data_iface: dummy
  hostname: data-node2
  data_iface_ip_addr: data-node2
  m0_clients:
    other: 0
    s3: 0
  memorysize_mb: 15866.52734375
  machine_id: bbb340f79047df9bb52fa460615c32a5
  processorcount: 8
  data_iface_type: tcp
- m0_servers:
  - io_disks:
      data:
      - path: /dev/sdc
        size: 53687091200
        blksize: 4096
      - path: /dev/sdd
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sdb'']'
    runs_confd: false
  - io_disks:
      data:
      - path: /dev/sdf
        size: 53687091200
        blksize: 4096
      - path: /dev/sdg
        size: 53687091200
        blksize: 4096
      meta_data: ! '[''/dev/sde'']'
    runs_confd: false
  - io_disks:
      data: []
      meta_data: null
    runs_confd: true
  data_iface: dummy
  hostname: data-node3
  data_iface_ip_addr: data-node3
  m0_clients:
    other: 0
    s3: 0
  memorysize_mb: 15866.52734375
  machine_id: ccc8700fe6797ed532e311b0615c32a7
  processorcount: 8
  data_iface_type: tcp
- m0_servers: null
  data_iface: dummy
  hostname: server-node1
  data_iface_ip_addr: server-node1
  m0_clients:
    other: 0
    s3: 2
  memorysize_mb: 15866.52734375
  machine_id: ddd120a9e051d103c164f605615c32a4
  processorcount: 8
  data_iface_type: tcp
- m0_servers: null
  data_iface: dummy
  hostname: server-node2
  data_iface_ip_addr: server-node2
  m0_clients:
    other: 0
    s3: 2
  memorysize_mb: 15866.52734375
  machine_id: eee340f79047df9bb52fa460615c32a5
  processorcount: 8
  data_iface_type: tcp
- m0_servers: null
  data_iface: dummy
  hostname: server-node3
  data_iface_ip_addr: server-node3
  m0_clients:
    other: 0
    s3: 2
  memorysize_mb: 15866.52734375
  machine_id: fff8700fe6797ed532e311b0615c32a7
  processorcount: 8
  data_iface_type: tcp
```

File: cfgen/cfgen (hunk 1 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: Done!

File: cfgen/cfgen (hunk 2 above)
Comment (suvratjoshi) [reply cfgen/cfgen:467]: Done!

File: provisioning/miniprov/hare_mp/utils.py [hunk 8]
```diff
@@ -98,28 +98,39 @@ def _save_drive_info(self, path: str):
         disk_key = path.strip('/')
         self.kv.kv_put(f'{hostname}/{disk_key}', drive_info)
 
+    def is_motr_component(self, machine_id: str) -> bool:
+        comp_names = self.provider.get(f'node>{machine_id}>'
```
[the remark this answers is not in the document]
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/utils.py:107]: Done!

File: cfgen/cfgen (hunk 3 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: Done!
As discussed with Mandar, this possibility need to be verified and and tested and improvement done using other JIRA.

File: cfgen/cfgen (hunk 4 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: Done!

File: hax/hax/queue/cli.py (hunk 5 above)
Comment (suvratjoshi) [reply hax/hax/queue/cli.py]: Yes this was rebase related issue.
Resolved!

File: provisioning/miniprov/hare_mp/main.py (hunk 6 above)
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/main.py]: Yes this was rebase related issue.

Resolved!

File: provisioning/miniprov/hare_mp/utils.py [hunk 9]
```diff
@@ -98,28 +98,44 @@ def _save_drive_info(self, path: str):
...
+        ConfStore (ValueProvider).
+        """
+        comp_names = self.provider.get(f'node>{machine_id}>'
+                                       f'components')
+        for component in comp_names:
+            if(component.get('name') == Const.COMPONENT_MOTR.value):
+                rc = True
+                break
+            else:
+                rc = False
```
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/utils.py:114]: It doesn't make sense to assign rc to False at every iteration.

File: provisioning/miniprov/hare_mp/utils.py [hunk 10]
```diff
@@ -98,28 +98,44 @@ def _save_drive_info(self, path: str):
...
+        """
+        Returns True if motr component is present in the components list
+        for the given node>{machine_id} according to the
+        ConfStore (ValueProvider).
+        """
+        comp_names = self.provider.get(f'node>{machine_id}>'
+                                       f'components')
+        for component in comp_names:
+            if(component.get('name') == Const.COMPONENT_MOTR.value):
+                rc = True
```
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/utils.py:111]: Why not like this
```suggestion
                return True
```
?

**Review by knekrasov — APPROVED** [review]
Looks ok, a couple of minor things to consider.

File: provisioning/miniprov/hare_mp/utils.py (hunk 10 above)
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/utils.py:111]: This is just by considering the generic rule that there should not be multiple return statements in any function. Then code becomes difficult to understand. Though this function is too small, but this is done for sticking to rule.

File: provisioning/miniprov/hare_mp/utils.py (hunk 9 above)
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/utils.py:114]: There can be sequence like this through which for loop will iterate through :
Example 1 : utils, motr, hare
Example 2 : utils, s3, hare
and change of sequences of component names in each of above examples.
This function is so written to work well in any such condition.

**Comment by suvratjoshi** [comment]
@mssawant ,
After addressing the review comments I created an image from custom build : 
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1579/
Ran deployment on 3 node setup and it was successful.
```
NAME                            READY   STATUS    RESTARTS   AGE
consul-7h2xs                    1/1     Running   0          18m
consul-mzmnf                    1/1     Running   0          18m
consul-server-0                 1/1     Running   0          18m
consul-server-1                 1/1     Running   0          18m
consul-server-2                 1/1     Running   0          18m
consul-tc79g                    1/1     Running   0          18m
control-node-b5ccb6877-j9dcl    4/4     Running   0          17m
data-node1-7dddb8fcc7-tnht6     4/4     Running   0          15m
data-node2-58b9b74d67-bth8c     4/4     Running   0          15m
data-node3-57c444bb6c-6btng     4/4     Running   0          15m
ha-node-7f5d45db58-qhkrs        1/1     Running   0          16m
kafka-0                         1/1     Running   0          18m
kafka-1                         1/1     Running   0          18m
kafka-2                         1/1     Running   0          18m
server-node1-68cfc6779f-9mjrg   6/6     Running   0          13m
server-node2-6589f65cd9-mjgvt   6/6     Running   0          13m
server-node3-df976d597-xx7x7    6/6     Running   0          13m
symas-openldap-pod              1/1     Running   0          18m
zookeeper-0                     1/1     Running   0          18m
zookeeper-1                     1/1     Running   0          18m
zookeeper-2                     1/1     Running   0          18m
```
hctl status :
```
[root@data-node1 /]# hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x90 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xdb 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    data-node2
    [started]  hax        0x7200000000000001:0x2b  inet:tcp:data-node2@2001
    [started]  ioservice  0x7200000000000001:0x2e  inet:tcp:data-node2@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:data-node2@3002
    [started]  confd      0x7200000000000001:0x48  inet:tcp:data-node2@3003
    data-node1
    [started]  hax        0x7200000000000001:0x7   inet:tcp:data-node1@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:data-node1@3001
    [started]  ioservice  0x7200000000000001:0x17  inet:tcp:data-node1@3002
    [started]  confd      0x7200000000000001:0x24  inet:tcp:data-node1@3003
    data-node3
    [started]  hax        0x7200000000000001:0x4f  inet:tcp:data-node3@2001
    [started]  ioservice  0x7200000000000001:0x52  inet:tcp:data-node3@3001
    [started]  ioservice  0x7200000000000001:0x5f  inet:tcp:data-node3@3002
    [started]  confd      0x7200000000000001:0x6c  inet:tcp:data-node3@3003
    server-node1
    [started]  hax        0x7200000000000001:0x71  inet:tcp:server-node1@2001
    [started]  s3server   0x7200000000000001:0x74  inet:tcp:server-node1@4001
    [started]  s3server   0x7200000000000001:0x77  inet:tcp:server-node1@4002
    server-node2
    [started]  hax        0x7200000000000001:0x7c  inet:tcp:server-node2@2001
    [started]  s3server   0x7200000000000001:0x7f  inet:tcp:server-node2@4001
    [started]  s3server   0x7200000000000001:0x82  inet:tcp:server-node2@4002
    server-node3  (RC)
    [started]  hax        0x7200000000000001:0x87  inet:tcp:server-node3@2001
    [started]  s3server   0x7200000000000001:0x8a  inet:tcp:server-node3@4001
    [started]  s3server   0x7200000000000001:0x8d  inet:tcp:server-node3@4002
[root@data-node1 /]#
```

**Comment by suvratjoshi** [comment]
The deployment for my PR https://github.com/Seagate/cortx-hare/pull/1911 using regular way of deployment has failed.
http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/599/console
 
It is failing in s3_setup config phase :
```
2021-12-23 18:23:46 cortx_setup [45]: INFO [_provision_components] /opt/seagate/cortx/s3/bin/s3_setup config --config yaml:///etc/cortx/cluster.conf --services io,auth,bg_consumer
2021-12-23 18:23:49 cortx_setup [45]: ERROR [main] error(22): config phase of s3, failed. b''
2021-12-23 18:23:49 cortx_setup [45]: ERROR [main] Traceback (most recent call last):
File "/usr/lib/python3.6/site-packages/cortx/setup/cortx_setup.py", line 132, in main
rc = command.process()
File "/usr/lib/python3.6/site-packages/cortx/setup/cortx_setup.py", line 120, in process
CortxProvisioner.cluster_bootstrap(self._args.cortx_conf, force_override)
File "/usr/lib/python3.6/site-packages/cortx/provisioner/provisioner.py", line 289, in cluster_bootstrap
CortxProvisioner._provision_components(cortx_conf, mini_prov_interfaces, apply_phase)
File "/usr/lib/python3.6/site-packages/cortx/provisioner/provisioner.py", line 251, in _provision_components
components[comp_idx]['name'], err)
cortx.provisioner.error.CortxProvisionerError: error(22): config phase of s3, failed. b''
```

**Review by SwapnilGaonkar7 — APPROVED** [review]

**Comment by suvratjoshi** [comment]
> @suvratjoshi, Idea is to have same CDF on all the nodes, I am confused by the statement `In any scenarios, config stage of hare should check, if motr services are there and then only generate motr related fids.`, does motr services also include s3server? So,
> 
> * CDF should be same on motr ioservices pod and s3server pods
> * consul-client-conf for s3server pod must contain only the s3server services.
> * while the motr ioservice pod will not have s3services present
>   If we take care of the 1, 2 and 3 must already be supported.

Hi @mssawant ,
Yes. We are taking that casre.
Now I have updated description.
Thanks!

File: provisioning/miniprov/hare_mp/cdf.py (hunk 7 above)
Comment (suvratjoshi) [reply provisioning/miniprov/hare_mp/cdf.py:356]: Now this patch is updated with latest commit where we are checking s3 component separately, to fix deployement in regular way - issue.
Thanks!

**Review by mssawant — APPROVED** [review]
Minor refactoring needed as discussed earlier, but let's do that as a separate PR.
